### PR TITLE
Add gpt2 architecture integration

### LINKED ISSUE #4073: GPT2 Architecture Integration
# Feature Description
The idea is to be able to convert models using the GPT2 architecture into GGUF. The convert-hf-to-gguf.py should include GPT2, as well as llama.cpp for running the model.

# Motivation
There are quite a few models for low resource languages or specific use cases that are fine-tuned on GPT2 architecture.

# Possible Implementation
The structure of models is quite similar to Starcoder. From my understanding, you can modify it quite easily by:

convert-hf-to-gguf.py
- Add a new model class
- Modify the set_gguf_parameters() [kv heads] and write_tensors() [maybe you need to transpose the qkv, up-ffn and down-ffn layer] methods

llama.cpp
- Add an new model class

# Status
I tried implementing that myself, but am not deep enough into the topic and find it quite hard to understand the libraries structure (is there any good documentation). So, I am probably not able to pull this off by myself, but am happy to support!

### PR DESCRIPTION
**Pull request by manikbhandari**
Closes #4073 

Output from GPT2-small:

`./main -p "Hi, I'm a language model," -n 20 -m models/models--gpt2/ggml-model-f16.gguf`

> Hi, I'm a language model, so please forgive me if my translation is bad.
> 
> Anyway, just in case it wasn't

### CODE REVIEW

File: convert-hf-to-gguf.py [hunk 1]
```diff
@@ -984,6 +988,66 @@ def write_tensors(self):
...
+        self.gguf_writer.add_embedding_length(self.hparams["n_embd"])
+        self.gguf_writer.add_feed_forward_length(4 * self.hparams["n_embd"])
+        self.gguf_writer.add_head_count(self.hparams["n_head"])
+        self.gguf_writer.add_layer_norm_eps(self.hparams["layer_norm_epsilon"])
+        self.gguf_writer.add_file_type(self.ftype)
+
+    def write_tensors(self):
+        block_count = self.hparams.get("n_layers", self.hparams.get("num_hidden_layers", self.hparams.get("n_layer")))
+        tensor_map = gguf.get_tensor_name_map(self.model_arch, block_count)
+        # import pdb; pdb.set_trace()
```
Comment (Galunid) [inline convert-hf-to-gguf.py:1005]: Not needed

File: llama.cpp [hunk 2]
```diff
@@ -5637,6 +5737,104 @@ struct llm_build_context {
 
         ggml_build_forward_expand(gf, cur);
 
+        return gf;
+    }
```
Comment (Galunid) [inline llama.cpp:5741]: Missing new line ;)

**Review by Galunid — COMMENTED** [review]
Please add `ggml-vocab-gpt2` test (run `python convert-hf-to-gguf.py --vocab-only ...`) and add the resulting file under `models` and relevant tests in `tests/CMakeLists.txt`. Similar to what was done in #3586

**Comment by manikbhandari** [comment]
Thanks for the review @Galunid ! 

- [x] Add GPT2 vocab file
- [x] Add test for GPT2 vocab
- [x] Remove debugging statement
- [x] Improve formatting in llama.cpp
- [x] Fix flake8 linting error

**Review by ggerganov — APPROVED** [review]
Need to merge `master` and adapt to the recent changes (`ggml_scale`, vram calc, etc.). We can merge after that

**Comment by manikbhandari** [comment]
Thanks for the review @ggerganov. Adapted the code with recent changes and added GPT-2 to readme.

File: llama.cpp [hunk 3]
```diff
@@ -5732,6 +5814,105 @@ struct llm_build_context {
...
+        cb(inpL, "inp_embd", -1);
+
+        // inp_pos - contains the positions
+        struct ggml_tensor * inp_pos = ggml_new_tensor_1d(ctx0, GGML_TYPE_I32, n_tokens);
+        cb(inp_pos, "inp_pos", -1);
+
+        // KQ_scale
+        struct ggml_tensor * KQ_scale = ggml_new_tensor_1d(ctx0, GGML_TYPE_F32, 1);
+        cb(KQ_scale, "KQ_scale", -1);
+
```
Comment (ggerganov) [inline llama.cpp:5835]: Not used - delete

**Review by ggerganov — APPROVED** [review]

**Review by Galunid — APPROVED** [review]
Once gg's suggestion is committed ;)

**Comment by manikbhandari** [comment]
Thanks! Removed the unused lines.

**Pull request merged**